Header: Use different templates based on page heirarchy

ryelle opened the pull request:
See https://github.com/WordPress/wporg-mu-plugins/pull/573 — There are updates to how the local navigation should appear across the site in an effort to simplify, make more responsive, and more consistent. The functionality changes are in `wporg-mu-plugins`, this companion PR serves to update the templates.

The header templates are included in the page templates, except for the new `header-third` template. Since this is included based on "hierarchy" rather than template-type, there's a filter in place to swap out the requested header templates. For example, the Block Handbook homepage and interior pages both use `single-handbook.html` templates, so the template calls for `header-alt`, but if the page is an interior page, it's replaced with `header-third`.

This PR also introduces a new block, `page-title`, which serves as a combination of query-title and post-title in one, using the post title on single posts and a search or archive title for those views. It also adds the `()` to relevant parser type titles. This avoids needing to create multiple `header-alt`/`header-third` templates for each type of content.

**Background**

There are now 3 header possibilities depending on page hierarchy, with 2 variations — in this PR, they correspond to:

| Design name | Template |
|---|---|
| Level 1A | `patterns/front-page-header.php` |
| Level 2A | `parts/header.html` |
| Level 2B | `parts/header-alt.html` |
| Level 3+ | `parts/header-third.html` |

We don't need to worry about Level 1B, because Level 1 only applies to home, so each site/section should only have one of the two.

Level 2 are items one level deep, with "2A" being things in the local nav, and "2B" not in the local nav. Splitting them avoids repeating the page title in the local nav bar.

Level 3+ are below that, and have breadcrumbs so you can jump back up to Level 2 & 1 (home).

**Screenshots**

Note: I used the WP-CLI command page to show L2A, but there's a bug that's causing it to not be highlighted in the local nav, tracked here: https://github.com/WordPress/wporg-developer/issues/498.

| Default | Scroll | Small screen |
|---|---|---|
| ![level-1](https://github.com/WordPress/wporg-developer/assets/541093/b62ab111-b002-40ca-826f-cfdafd1180d4) | ![level-1-scroll](https://github.com/WordPress/wporg-developer/assets/541093/758801f5-7983-4130-a11b-a507565cdb7f) | ![level-1-mobile](https://github.com/WordPress/wporg-developer/assets/541093/37a5109f-ee6e-4068-b1fc-50f258c51e3c) |
| ![level-2a](https://github.com/WordPress/wporg-developer/assets/541093/2a96ff69-fde3-40ee-a03a-ea3f4284e5db) | ![level-2a-scroll](https://github.com/WordPress/wporg-developer/assets/541093/85699aee-4ff6-473e-8e43-065634a26643) | ![level-2a-mobile](https://github.com/WordPress/wporg-developer/assets/541093/4bf7e5e2-7a90-4842-ae59-c4463a3450e8) |
| ![level-2b](https://github.com/WordPress/wporg-developer/assets/541093/3d7cee4e-280a-41a0-94d6-50df3fe211c4) | ![level-2b-scroll](https://github.com/WordPress/wporg-developer/assets/541093/2017543f-90e4-42f7-8605-f224d0f874a5) | ![level-2b-mobile](https://github.com/WordPress/wporg-developer/assets/541093/984ff353-6469-4708-9962-54f3c719ba2d) |
| ![level-3](https://github.com/WordPress/wporg-developer/assets/541093/eb342287-f1f0-4caa-a063-999c156205a4) | ![level-3-scroll](https://github.com/WordPress/wporg-developer/assets/541093/6194bc2d-ab8d-4ce7-9b7b-14a36327734c) | ![level-3-mobile](https://github.com/WordPress/wporg-developer/assets/541093/5b583cdf-5672-45ad-a481-a8a704fe610e) |

To test:

For the best look, apply https://github.com/WordPress/wporg-mu-plugins/pull/573 to wporg-mu-plugins.

View the following pages, and make sure they're using the expected template.

- L1A: https://developer.wordpress.org/
- L2A: https://developer.wordpress.org/cli/commands/
- L2A: https://developer.wordpress.org/reference/
- L2B: https://developer.wordpress.org/block-editor/
- L3+: https://developer.wordpress.org/reference/functions/get_stylesheet_directory/
- L3+: https://developer.wordpress.org/cli/commands/cache/
- L3+: https://developer.wordpress.org/block-editor/how-to-guides/data-basics/2-building-a-list-of-pages/

ryelle committed:
Update homepage header

ryelle committed:
Block Hooks: Update template part depending on the current page.

ryelle committed:
Add the "Level 2 B" header variation as "header-alt"

ryelle committed:
New block: Add a page title block which can be used for single, archive, and search page titles.

ryelle committed:
Unhide breadcrumbs on small screens

adamwoodnz commented:
The top of the search results page has become pretty tight now that breadcrumbs are not displayed, should probably be a 20px gap at the top, or maybe 40px.

| Prod | Local |
|-|-|
| ![developer wordpress org__s=block(3 1536x864)](https://github.com/WordPress/wporg-developer/assets/1017872/7889cb86-6c44-45bf-8a94-a6279371fbf8) | ![localhost_8888__s=block(3 1536x864)](https://github.com/WordPress/wporg-developer/assets/1017872/bb841a89-692c-4679-bed9-cf29d90b404f) |

adamwoodnz commented:
On top level handbook pages there's an issue with misaligned sidebar containers now that the breadcrumbs aren't displayed:

https://github.com/WordPress/wporg-developer/assets/1017872/af93572c-1fba-428b-8ab6-4ce144133889

I think we'll need to change the `--wp--custom--wporg-sidebar-container--spacing--margin--top` var for those pages, as that's what controls the top margin of the ToC. `100px` works for me.

adamwoodnz reviewed (changes requested):
👍 working well for me apart from a couple of things inline

outdoor2kode reviewed (commented):
Have you encountered this situation? At `/apis`.

https://github.com/WordPress/wporg-developer/assets/18050944/b2084390-cf88-45fc-8614-f6c67e75d7d5

ryelle committed:
Fade in page title on scroll

ryelle committed:
Fix spacing of Table of Contents sidebar

ryelle committed:
Update space over search box to match other pages

ryelle commented:
> On top level handbook pages there's an issue with misaligned sidebar containers now that the breadcrumbs aren't displayed

Fixed in https://github.com/WordPress/wporg-developer/pull/500/commits/9eaf52eaca3dbec7c4c203d915615387fda7be49

![Screenshot 2024-02-16 at 6 17 45 PM](https://github.com/WordPress/wporg-developer/assets/541093/14a61353-5f2a-4435-a68d-4620336b2afa)

> The top of the search results page has become pretty tight now that breadcrumbs are not displayed

I've updated this to match the other pages, so it's 20px now.  https://github.com/WordPress/wporg-developer/pull/500/commits/f018c3ffe6070d7e7f7c4ede89541d906e42f75c

![Screenshot 2024-02-16 at 6 17 31 PM](https://github.com/WordPress/wporg-developer/assets/541093/ea010880-591c-4a93-a7b8-d180908c3791)

> Have you encountered this situation? At /apis.

@renintw I have, but I don't think it's related to this PR, I can replicate it on production.

outdoor2kode reviewed (approved):
LGTM 👍 

> I have, but I don't think it's related to this PR, I can replicate it on production.

Thanks. I've opened a [ticket](https://github.com/WordPress/wporg-developer/issues/503) for it.

adamwoodnz commented on source/wp-content/themes/wporg-developer-2023/theme.json:
```diff
@@ -158,7 +158,7 @@
 				"spacing": {
 					"margin": {
-						"top": "150px"
+						"top": "100px"
```
Sorry, obviously more clarity required; on pages with breadcrumbs it needs to remain 150px, but for the top level handbook pages without breadcrumbs 100px.

The child page sidebars are now misaligned:

![Screenshot 2024-02-20 at 12 13 47 PM](https://github.com/WordPress/wporg-developer/assets/1017872/9e21f184-7e7f-49c8-a130-d7f96f9c5126)

ryelle committed:
Fix sidebar alignment on landing & inner handbook pages

ryelle commented:
Okay, I think I have it now @adamwoodnz — how's it looking to you?

<img width="1396" alt="Screenshot 2024-02-21 at 4 55 22 PM" src="https://github.com/WordPress/wporg-developer/assets/541093/f0751d28-2aeb-4fe8-9834-50e99e0fbdfd">
<img width="1398" alt="Screenshot 2024-02-21 at 4 55 14 PM" src="https://github.com/WordPress/wporg-developer/assets/541093/2961347c-2094-48db-8046-a351535868ee">

ryelle commented:
Rebased this PR to make sure it still works after the updated layout changes 👍🏻

adamwoodnz commented:
> Okay, I think I have it now @adamwoodnz — how's it looking to you?

LGTM!
